[Blog] Vectors, Matrices & Matrix Order Blog (#2212)
o3de/o3de.org: content/blog/posts/vectors-matrices-matrix-order.md

---
title: "Vectors, Matrices & Matrix Order"
date: 2023-02-22
slug: vectors-matrices-matrix-order
author: John Coimbra Walsh
blog_img: "/images/blog/dev-series.JPG"
full_img: ""
---

In this post I will attempt to demystify a topic that is the source of confusion for a lot of people as they dip their toes into learning the concepts that underpin computer graphics: vectors, matrices and matrix order.

## Matrices & vectors

An \\(n \cdot m\\) matrix consists of \\(n\\) rows and \\(m\\) columns. The product of two matrices \\(\mathbf{A}\\) and \\(\mathbf{B}\\) is well defined when the number of columns in \\(\mathbf{A}\\) matches the number of rows in \\(\mathbf{B}\\). Their product is a matrix with the number of rows of \\(\mathbf{A}\\) and the number of columns of \\(\mathbf{B}\\). Thus, for a given \\(2×3\\) matrix \\(\mathbf{A}\\) to be multiplied with a given \\(3×2\\) matrix \\(\mathbf{B}\\), the product is calculated as follows to produce the resulting \\(2×2\\) matrix:

$$
\begin{array}{l@{}l}
\mathbf{A} \cdot \mathbf{B} 

>     &{}= \begin{bmatrix} a_{11} & a_{12} & a_{13} \\\ a_{21} & a_{22} & a_{23} \end{bmatrix} \cdot \begin{bmatrix} b_{11} & b_{12} \\\ b_{13} & b_{21} \\\ b_{22} & b_{23} \end{bmatrix} \\\ \\\

Reviewer: ```suggestion
    &{}= \begin{bmatrix} a_{11} & a_{12} & a_{13} \\\ a_{21} & a_{22} & a_{23} \end{bmatrix} \cdot \begin{bmatrix} b_{11} & b_{12} \\ b_{21} & b_{22} \\ b_{31} & b_{32} \end{bmatrix} \\\ \\\
```

a_{21}b_{11} + a_{22}b_{21} + a_{23}b_{31} & a_{21}b_{12} + a_{22}b_{22} + a_{23}b_{32} \end{bmatrix}
\end{array}
$$

Vectors can be written as either rows:

$$\vec{r} = \begin{bmatrix} r_{1} & r_{2} & r_{3} & ... & r_{m}  \end{bmatrix}$$

Or columns:

$$\vec{c} = \begin{bmatrix} c_{1} \\\ c_{2} \\\ c_{3} \\\ ... \\\ c_{m}  \end{bmatrix}$$

The vector-matrix product follows the same rules as the matrix product, albeit treating row vectors as \\(1×m\\) matrices and column vectors as \\(m×1\\) matrices. Thus, for a given \\(2×2\\) matrix \\(\mathbf{M}\\) to be multiplied with a given \\(1×2\\) row vector \\(\vec{r}\\), we must ensure that the number of columns of the matrix on the left hand side matches the number of rows of the matrix on the right hand side. We achieve this by pre-multiplying the vector \\(\vec{r}\\) with \\(\mathbf{M}\\), resulting in a \\(1×2\\) matrix:

$$
\begin{array}{l@{}l}
\vec{r} \cdot \mathbf{M} 
    &{}= \begin{bmatrix} r_{1} & r_{2} \end{bmatrix} \cdot \begin{bmatrix} m_{11} & m_{12} \\\ m_{21} & m_{22}  \end{bmatrix} \\\ \\\
    &{}= \begin{bmatrix} 
r_{1}m_{11} + r_{2}m_{21} & 
r_{1}m_{12} + r_{2}m_{22} \end{bmatrix}
\end{array}
$$

> In order to multiply a column vector by the same matrix, we must ensure that the number of columns of the matrix on the left hand side matches the number of rows of the matrix on the right hand side. We achieve this by post-multiplying the vector \\(\vec{c}\\) with \\(\mathbf{M}\\), resulting in a \\(2×1\\) matrix:

Reviewer 2: I'd say it's a little confusing to call it the same matrix. It is the same matrix, but it will be a different transformation when you work with a column vector vs a row vector.

Reviewer: ```suggestion
In order to multiply a column vector, we must ensure that the number of columns of the matrix on the left hand side matches the number of rows of the matrix on the right hand side. We achieve this by post-multiplying the vector \\(\vec{c}\\) with \\(\mathbf{M}\\), resulting in a \\(2×1\\) matrix:
```

$$
\begin{array}{l@{}l}
\mathbf{M} \cdot \vec{c} 
    &{}= \begin{bmatrix} m_{11} & m_{12} \\\ m_{21} & m_{22}  \end{bmatrix} \cdot \begin{bmatrix} c_{1} \\\ c_{2} \end{bmatrix} \\\ \\\
&{}= \begin{bmatrix} 

> m_{11}c_{1} + m_{21}c_{1} \\\ 

Reviewer 2: I think this should be m_{11}c_{1} + m_{12}c_{2} \\\

Reviewer: ```suggestion
m_{11}c_{1} + m_{12}c_{2} \\
```

> m_{12}c_{2} + m_{22}c_{2} \end{bmatrix}

Reviewer 2: m_{21}c_{1} + m_{22}c_{2} \end{bmatrix}

Reviewer: ```suggestion
m_{21}c_{1} + m_{22}c_{2} \end{bmatrix}
```

\end{array}
$$

> However, due to the non-commutativity of matrix multiplication, $\vec{r} \cdot \mathbf{M}$ may not equal to $\mathbf{M} \cdot \vec{c}$ as the two products are necessarily not equivalent:

Reviewer 2: This isn't really related to commutativity. Matrix non-commutativity means that AB does not in general equal BA. But r and c are not the same matrix, they are each other's transpose. The relevant fact is that (AB)ᵀ = BᵀAᵀ.

Reviewer: We discussed offline and I am making the suggested phrasing as suggested edits 👍

> However, due to the non-commutativity of matrix multiplication, $\vec{r} \cdot \mathbf{M}$ may not equal to $\mathbf{M} \cdot \vec{c}$ as the two products are necessarily not equivalent:

Reviewer: ```suggestion
Since \\(\vec{c}\\) is the transpose of \\(\vec{r}\\), and transpose multiplication follows the identity \\((\mathbf{A}\mathbf{B})^\mathbf{T} = \mathbf{B^T}\mathbf{A^T}\\), we need to transpose \\(\mathbf{M}\\) and reverse the order in the multiplication to get the equivalent product as a column vector. If we didn't reverse the order and take the transpose, we would get a different result:
```

$$
 \begin{bmatrix} 1 & 0 & 1 \end{bmatrix} \cdot \begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix}     \neq \begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix}
$$

 The row vector product is as follows:

$$
\begin{bmatrix} 1 & 0 & 1 \end{bmatrix} \cdot \begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix} = \begin{bmatrix} 4 & 0 & 4 \end{bmatrix}
$$

Whereas the column vector product is as follows:

$$

> \begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 2 & 0 & 6 \end{bmatrix}

Reviewer: Ah, my bad! will fix and resolve the other comments 👍

> \begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 2 & 0 & 6 \end{bmatrix}

Reviewer: ```suggestion
\begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 2 \\ 0 \\ 6 \end{bmatrix}
```

$$

> Furthermore, the dimensions of \\(\mathbf{M}\\) may be well defined for \\(\vec{r} \cdot \mathbf{M}\\) but not necessarily for \\(\mathbf{M} \cdot \vec{c}\\). Thus, in order to make \\(\mathbf{M} \cdot \vec{c}\\) well-defined and equivalent to \\(\vec{r} \cdot \mathbf{M}\\), we must multiply \\(\vec{c}\\) by the transpose of \\(\mathbf{M}\\), such that \\(\vec{r} \cdot \mathbf{M} = \mathbf{M^T} \cdot \vec{c}\\) (and visa versa):

Reviewer: ```suggestion
Another way of thinking about this is that in order to have a well-defined product, the row vector has to appear first in the product, and since we have transposed the vector, we also need to transpose the matrix so for \\(\vec{c}\\) we would have:
```

$$
\begin{array}{l@{}l}
\mathbf{M}^{T} \cdot \vec{c} 
    &{}= \begin{bmatrix} m_{11} & m_{21} \\\ m_{12} & m_{22}  \end{bmatrix} \cdot \begin{bmatrix} c_{1} \\\ c_{2} \end{bmatrix} \\\ \\\
&{}= \begin{bmatrix} 
m_{11}c_{1} + m_{12}c_{1} \\\ 
m_{21}c_{2} + m_{22}c_{2} \end{bmatrix}
\end{array}
$$

With this in mind, the column vector product between the transpose of the matrix used in the vector product is now equal to that of the vector product as follows:

$$

> \begin{bmatrix} 1 & 0 & 3 \\\ 0 & 2 & 0 \\\ 1 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 4 & 0 & 4 \end{bmatrix}

Reviewer 2: rhs should be a column vector again

> \begin{bmatrix} 1 & 0 & 3 \\\ 0 & 2 & 0 \\\ 1 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 4 & 0 & 4 \end{bmatrix}

Reviewer: ```suggestion
\begin{bmatrix} 1 & 0 & 3 \\\ 0 & 2 & 0 \\\ 1 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 4 \\ 0 \\ 4 \end{bmatrix}
```

$$

> The matrices themselves have no concept of “row vectors” or “column vectors”, from the perspective of matrix products there’s only matrices with the same well defined rules applied.

Reviewer 2: Could be worth saying there is a difference in that it's a 1x3 matrix or a 3x1 matrix.

$$
\begin{bmatrix} a & b & c & d \\\ e & f & g & h \\\ i & j & k & l \end{bmatrix}
$$

We can interpret two dimensional arrays (or flat one dimensional arrays of size \\(m \cdot n\\), the distinction isn’t important) as tabular data consisting of a number of rows and columns. This lends itself well to the implementation of matrices in C++ as we can then use our major and minor indices to access the specific elements of our matrix. However, without any context, accessing the elements in the array representation of a matrix using the major and minor indices is ambiguous. Consider the matrix above as represented by a two dimensional array. What will the output of the code snippet below be?

```c++
for(std::size_t j = 0; j < m; j++)
{
   for(std::size_t i = 0; i < n; i++)
   {
      std::cout << matrix[j][i] << " ";
   }
   
   std::cout << "\n";
}
```

There are two possible answers, depending on whether we treat each one dimensional array (accessed by the major index) as a *row* or as a *column* within the matrix:

<table>
<tr>
<th> Major index as rows </th> <th> Major index as columns </th>
</tr>
<tr>
<td>
<pre lang="c++">
a b c d
e f g h
i j k l
</pre>
</td>
<td>
<pre lang="c++">

> a e j

Reviewer 2: think this should be a e i

Reviewer: ```suggestion
a e i
```

## Row & column major matrix representations

We can store the matrix above in row-major ordering as follows:

```c++
const std::size_t rows = 3;
const std::size_t cols = 4;

const char rowMajor[rows][cols] = 
{ 
    {'a', 'b', 'c', 'd'}, // row 0 
    {'e', 'f', 'g', 'h'}, // row 1
    {'i', 'j', 'k', 'l'}  // row 2
};

// prints the array representation of the matrix (which is also the actual matrix)
// a b c d
// e f g h
// i j k l
for(std::size_t r = 0; r < rows; r++)
{
   for(std::size_t c = 0; c < cols; c++)
   {
      std::cout << rowMajor[r][c] << " ";
   }
   
   std::cout << "\n";
}
```

Likewise, we can store our matrix in column-major ordering simply by swapping the dimensions and ordering of the data we initialize the array with:

```c++
const std::size_t rows = 3;
const std::size_t cols = 4;

const char colMajor[cols][rows] = 
{
    {'a', 'e', 'i'}, // column 0
    {'b', 'f', 'j'}, // column 1
    {'c', 'g', 'k'}, // column 2

>     {'d', 'h', 'l'}  // column 4

Reviewer: ```suggestion
    {'d', 'h', 'l'}  // column 3
```

      result[c] += rowMajor[r][c] * rowVector[r];
   }
}

// prints 4 0 4 8
for(std::size_t i = 0; i < cols; i++)
{
   std::cout << result[i] << " ";
}
```

Now consider the following product between the transpose of the above matrix and a column vector:

$$
\begin{bmatrix} 1 & 0 & 3 \\\ 0 & 2 & 0 \\\ 1 & 0 & 3 \\\ 4 & 1 & 4  \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} = \begin{bmatrix} 4 \\\ 0 \\\ 4 \\\ 8 \end{bmatrix}
$$

We can implement the matrix-vector product with column major array storage as follows:

```c++
// 3x1 column vector
const float columnVector[3] = {1.f, 0.f, 1.f};

// 4x3 matrix in column-major order
const std::size_t rows = 4;
const std::size_t cols = 3;

const float columnMajor[cols][rows] = 
{ 
    {1.f, 0.f, 1.f, 4.f}, // column 0 
    {0.f, 2.f, 0.f, 1.f}, // column 1
    {3.f, 0.f, 3.f, 4.f}  // column 2
};

// 4x1 column vector to hold the result
float result[rows] = {0.f, 0.f, 0.f, 0.f};

for(std::size_t r = 0; r < rows; r++)
{
   for(std::size_t c = 0; c < cols; c++)
   {
      result[r] += columnMajor[c][r] * columnVector[c];
   }
}

// prints 4 0 4 8
for(std::size_t i = 0; i < rows; i++)
{
   std::cout << result[i] << " ";
}
```

> You may have noticed that the code for both snippets is identical, bar some flipping of variable names for clarity. Remember that the product of a row vector with a given matrix is the same as the product of a column vector of the transpose of that same matrix:

Reviewer: ```suggestion
You may have noticed that the code for both snippets is identical, bar some flipping of variable names for clarity. Remember that the product of a row vector with a given matrix is the transpose of the product between a column vector and the transpose of that same matrix:
```

$$

> \vec{r} \cdot \mathbf{M} =\mathbf{M}^{T} \cdot \vec{c} 

Reviewer: ```suggestion
\vec{r} \cdot \mathbf{M} =\mathbf{M^T} \cdot \vec{c} 
```

Reviewer 2: There rhs is equal to the transpose of the lhs, not the lhs itself.

Reviewer: ```suggestion
\vec{r} \cdot \mathbf{M} = (\mathbf{M^T} \cdot \vec{c})^\mathbf{T}}
```

$$

An interesting property of storing our matrices in the array matrix order convention of our vector convention (i.e. row vectors with matrices stored in row-major order and column vectors with matrices stored in column-major order) is that the actual array data for the matrices is identical regardless of which of the two convention combinations you choose. Consider the following 2d transform matrix for scaling, anti-clockwise rotating and translating a given row vector:

$$

> \begin{bmatrix}      S_{x} \cdot \cos(\theta) & \sin(\theta) & 0  \\\      -\sin(\theta) & S_{y} \cdot \cos(\theta) & 0  \\\  T_{x} &  T_{y}  & 1  \end{bmatrix}

Reviewer: I don't follow... can you make a suggested edit for review?

$$

When laid out in row-major in a two dimensional array, the array is initialized as follows:

```c++
// Scale to double and quadruple the x and y axis accordingly
const float Sx = 2.f;
const float Sy = 4.f

// Rotation of approx. 45 degrees (in radians)

> const float theta = 0.707;

Reviewer 2: This looks like cos theta rather than theta. 45 degrees in radians would be about 0.785.

Reviewer: ```suggestion
const float theta = 0.785;
```

// Translation of -5 along x axis and 10 along y axis
const float Tx = -5.f;
const float Ty = 10.f;

// 3x3 matrix in row-major order
const std::size_t rows = 3;
const std::size_t cols = 3;

const float rowMajor[rows][cols] = 
{ 
    {Sx * cos(theta), sin(theta),      0.f}, // row 0 
    {-sin(theta),     Sy * cos(theta), 0.f}, // row 1
    {Tx,              Ty,              1.f}  // row 2
};
```

Now consider the same transformation matrix transposed for translating a given column vector:

$$
\begin{bmatrix}      S_{x} \cdot \cos(\theta) & -\sin(\theta) & T_{x} \\\      \sin(\theta) & S_{y} \cdot \cos(\theta) & T_{y} \\\      0 & 0 & 1  \end{bmatrix}
$$

When laid out in column-major in a two dimensional array, the array is initialized as follows:

```c++
// Scale to double and quadruple the x and y axis accordingly
const float Sx = 2.f;
const float Sy = 4.f

// Rotation of approx. 45 degrees (in radians)

> const float theta = 0.707;

Reviewer: ```suggestion
const float theta = 0.785;
```

// Translation of -5 along x axis and 10 along y axis
const float Tx = -5.f;
const float Ty = 10.f;

// 3x3 matrix in column-major order
const std::size_t rows = 3;
const std::size_t cols = 3;

const float columnMajor[cols][rows] = 
{ 
    {Sx * cos(theta), sin(theta),      0.f}, // column 0 
    {-sin(theta),     Sy * cos(theta), 0.f}, // column 1
    {Tx,              Ty,              1.f}  // column 2
};
```

This is convenient because if we wish to support both row vectors and column vectors, our utility functions to generate the various transformation matrices needed will be identical so long as we assume our matrix order conventions match our vector conventions. As to which combination of conventions is superior, it really is a matter of taste. Row vectors with row-major matrix storage have the slight advantage of the code layout mapping nicely to the row-by-row layout of matrix notation, whereas for column vectors with column-major matrix storage you need to mentally transpose the matrices when reading code. Ultimately, the only thing that is important is that the conventions you pick are documented and used consistently.

## Programming languages & matrix order

> Programming languages themselves have assumptions about the matrix order of multidimensional arrays that is independent of the assumptions about the matrix order of our array layout when used to represent matrices. For example, the C99 language specification §6.5.2.1p3 (thus, by extension, C++) states (emphasis mine):
> 
> _Successive subscript operators designate an element of a multidimensional array object. If E is an n-dimensional array (n >= 2) with dimensions i x j x . . . x k, then E (used as other than an lvalue) is converted to a pointer to an (n - 1)-dimensional array with dimensions j x . . . x k. If the unary * operator is applied to this pointer explicitly, or implicitly as a result of subscripting, the result is the referenced (n - 1)-dimensional array, which itself is converted into a pointer if used as other than an lvalue. **It follows from this that arrays are stored in row-major order (last subscript varies fastest).**_

Reviewer 3: ```suggestion
Programming languages themselves have assumptions about the matrix order of multidimensional arrays that is independent of the assumptions about the matrix order of our array layout when used to represent matrices. For example, the [C99 language specification §6.5.2.1p3](https://www.open-std.org/jtc1/sc22/WG14/www/docs/n1256.pdf) (thus, by extension, C++) states (emphasis added):

> Successive subscript operators designate an element of a multidimensional array object. If E is an n-dimensional array (n >= 2) with dimensions i x j x . . . x k, then E (used as other than an lvalue) is converted to a pointer to an (n - 1)-dimensional array with dimensions j x . . . x k. If the unary * operator is applied to this pointer explicitly, or implicitly as a result of subscripting, the result is the referenced (n - 1)-dimensional array, which itself is converted into a pointer if used as other than an lvalue. **It follows from this that arrays are stored in row-major order (last subscript varies fastest).**
```

Change to use block-quote style and link the reference.

> If we were to store our matrices in the natural matrix order of C++ but access the elements in opposing matrix order, we would have element access patterns that are cache unfriendly as each element access will not being contiguous but instead separated by a stride equal to the number of columns in the matrix. Consider the following:

Reviewer: ```suggestion
If we were to store our matrices in the natural matrix order of C++ but access the elements in opposing matrix order, we would have element access patterns that are cache unfriendly as each element access will not be contiguous but instead separated by a stride equal to the number of columns in the matrix. Consider the following:
```

As we have seen in the previous section, there is nothing stopping us from storing our matrices using a column-major layout if we wish to access our elements column-by-column rather than row-by-row so ideally we would pick conventions that match closest to our expected element access patterns. For transformation matrices, this question has a relatively straightforward answer: as our matrices will contain the basis vectors for our vector transformations, when we pick the matrix order convention that matches our vector convention, the basis vectors in our transformation matrices will be guaranteed to have contiguous components. On the flipside, if we don’t match our matrix order and vector conventions, our basis vector components will be separated in memory by a stride matching the matrix dimension of the matrix order  convention used to store the matrix. Consider the following example of retrieving the x-axis basis vector of a 2d rotation matrix for column-vectors stored in row-major:

```c++

// Rotation of approx. 90 degrees (in radians)

> const float theta = 0.707;

Reviewer: ```suggestion
const float theta = 0.785;
```

> // Rotation of approx. 90 degrees (in radians)

Reviewer: ```suggestion
// Rotation of approx. 45 degrees (in radians)
```

// 2x2 matrix in row-major order
const std::size_t rows = 2;
const std::size_t cols = 2;
using row_type = std::array<float, cols>;
using col_type = std::array<float, rows>;
using mat2x2_type = std::array<row_type, rows>;

// 2d rotation matrix for column vectors
const mat2x2_type rowMajor = 
{
    {cos(theta), -sin(theta)}, // row 0 
    {sin(theta),  cos(theta)}  // row 1
};

// returns the x basis vector
col_type basisX()
{
     // x-axis basis vector components stored non-contiguously
    return col_type 
    {
        rowMajor[0][0],
        rowMajor[1][0]
    };
}
```

Notice that we have to construct an intermediary array and populate it with the basis vector components due to the fact that they’re not stored contiguously in the array representation of the matrix. Now compare that to the following example of retrieving the x-axis basis vector of the same matrix stored in column-major:

```c++
// Rotation of approx. 90 degrees (in radians)

> const float theta = 0.707;

Reviewer 2: Same comment

Reviewer: ```suggestion
const float theta = 0.785;
```

> // Rotation of approx. 90 degrees (in radians)

Reviewer: ```suggestion
// Rotation of approx. 45 degrees (in radians)
```

// 2x2 matrix in column-major order
const std::size_t rows = 2;
const std::size_t cols = 2;
using col_type = std::array<float, rows>;
using mat2x2_type = std::array<col_type, cols>;

// 2d rotation matrix for column vectors
const mat2x2_type columnMajor = 
{ 

>     {Sx * cos(theta), sin(theta)}, // column 0 

Reviewer 2: Should this Sx be here? It hasn't been declared above and I don't think it's necessary for this example.

Reviewer: ```suggestion
    {cos(theta), sin(theta)}, // column 0 
```

    {-sin(theta),     cos(theta)}  // column 1
};

// returns the x basis vector
const col_type& basisX()
{
     // the x-axis basis vector components stored contiguously
    return columnMajor[0];
}
```

In the above example, rather than having to construct an intermediary array for our basis vector, we can simply return a reference to the first column in the matrix. When composing our vectors and matrices from more complex types, this becomes a useful property as it makes accessing the vectors that make up our transformation matrices a trivial operation without the need for constructing any intermediary types.

## Closing thoughts

When picking conventions for our vectors and matrices, we can choose any combination of row or column vectors with row- or column-major array representations of our matrices as these two conventions are independent of one another. However, not all choices are equal so I will present an (admittedly, subjective) summary of the pros and cons of each combination:

<table>
<tr>
<th>Convention</th> <th>Pros</th> <th>Cons</th>
</tr>

<tr>
<td>Row vectors with matrices stored in row-major order</td>
<td>
<ul>

> <li>Requires no reordering of array data for the matrices.</li>

Reviewer: ```suggestion
<li>Uses the same vector convention (and thus ordering of vector and matrix products) as a lot of graphics literature.</li>
<li>Requires no reordering of array data for the matrices.</li>
```

<li>Initialization of matrix array elements in code maps naturally to the layout of the matrix.</li>
<li>Cache-friendly, trivial access to the basis vectors in transformation matrices.</li>
</ul>
</td>
<td>
<ul>

> <li>Uses the opposite vector convention (and thus ordering of vector and matrix products) as mathematical literature.</li>

Reviewer 2: I think a lot of graphics textbooks will use row vectors though

---
title: "Vectors, Matrices & Matrix Order"
date: 2023-02-22
slug: vectors-matrices-matrix-order
author: John Coimbra Walsh
blog_img: "/images/blog/dev-series.JPG"
full_img: ""
---

> _In this blog post, AWS engineer John Coimbra Walsh covers the basics of using vectors, matricies, and matrix order. This is a complicated topic for novice graphics programmers, and hopefully this breakdown can help you!_

Reviewer: ```suggestion
_In this blog post, AWS engineer John Coimbra Walsh covers the basics of using vectors, matrices, and matrix order. This is a complicated topic for novice graphics programmers, and hopefully this breakdown can help you!_
```

## Matrices & vectors

An \\(n \cdot m\\) matrix consists of \\(n\\) rows and \\(m\\) columns. The product of two matrices \\(\mathbf{A}\\) and \\(\mathbf{B}\\) is well defined when the number of columns in \\(\mathbf{A}\\) matches the number of rows in \\(\mathbf{B}\\). Their product is a matrix with the number of rows of \\(\mathbf{A}\\) and the number of columns of \\(\mathbf{B}\\). Thus, for a given \\(2×3\\) matrix \\(\mathbf{A}\\) to be multiplied with a given \\(3×2\\) matrix \\(\mathbf{B}\\), the product is calculated as follows to produce the resulting \\(2×2\\) matrix:

$$
\begin{array}{l@{}l}
\mathbf{A} \cdot \mathbf{B} 

>     &{}= \begin{bmatrix} a_{11} & a_{12} & a_{13} \\\ a_{21} & a_{22} & a_{23} \end{bmatrix} \cdot \begin{bmatrix} b_{11} & b_{12} \\ b_{21} & b_{22} \\ b_{31} & b_{32} \end{bmatrix} \\\ \\\

Reviewer: ```suggestion
    &{}= \begin{bmatrix} a_{11} & a_{12} & a_{13} \\\ a_{21} & a_{22} & a_{23} \end{bmatrix} \cdot \begin{bmatrix} b_{11} & b_{12} \\ b_{21} & b_{22} \\ b_{31} & b_{32} \end{bmatrix} \\ \\
```

>     &{}= \begin{bmatrix} a_{11}b_{11} + a_{12}b_{21} + a_{13}b_{31} & a_{11}b_{12} + a_{12}b_{22} + a_{13}b_{32} \\\ 

Reviewer: ```suggestion
    &{}= \begin{bmatrix} a_{11}b_{11} + a_{12}b_{21} + a_{13}b_{31} & a_{11}b_{12} + a_{12}b_{22} + a_{13}b_{32} \\ 
```

a_{21}b_{11} + a_{22}b_{21} + a_{23}b_{31} & a_{21}b_{12} + a_{22}b_{22} + a_{23}b_{32} \end{bmatrix}
\end{array}
$$

Vectors can be written as either rows:

$$\vec{r} = \begin{bmatrix} r_{1} & r_{2} & r_{3} & ... & r_{m}  \end{bmatrix}$$

Or columns:

$$\vec{c} = \begin{bmatrix} c_{1} \\\ c_{2} \\\ c_{3} \\\ ... \\\ c_{m}  \end{bmatrix}$$

> The vector-matrix product follows the same rules as the matrix product, albeit treating row vectors as \\(1×m\\) matrices and column vectors as \\(m×1\\) matrices. Thus, for a given \\(2×2\\) matrix \\(\mathbf{M}\\) to be multiplied with a given \\(1×2\\) row vector \\(\vec{r}\\), we must ensure that the number of columns of the matrix on the left hand side matches the number of rows of the matrix on the right hand side. We achieve this by pre-multiplying the vector \\(\vec{r}\\) with \\(\mathbf{M}\\), resulting in a \\(1×2\\) matrix:

Reviewer: ```suggestion
The vector-matrix product follows the same rules as the matrix product, albeit treating row vectors as \\(1×m\\) matrices and column vectors as \\(m×1\\) matrices. Thus, for a given \\(2×2\\) matrix \\(\mathbf{M}\\) to be multiplied with a given \\(1×2\\) row vector \\(\vec{r}\\), we must ensure that the number of columns of the matrix on the left hand side matches the number of rows of the matrix on the right hand side. We achieve this by post-multiplying the vector \\(\vec{r}\\) by \\(\mathbf{M}\\), resulting in a \\(1×2\\) matrix:
```

$$
\begin{array}{l@{}l}
\vec{r} \cdot \mathbf{M} 
    &{}= \begin{bmatrix} r_{1} & r_{2} \end{bmatrix} \cdot \begin{bmatrix} m_{11} & m_{12} \\\ m_{21} & m_{22}  \end{bmatrix} \\\ \\\
    &{}= \begin{bmatrix} 
r_{1}m_{11} + r_{2}m_{21} & 
r_{1}m_{12} + r_{2}m_{22} \end{bmatrix}
\end{array}
$$

> In order to multiply a column vector, we must ensure that the number of columns of the matrix on the left hand side matches the number of rows of the matrix on the right hand side. We achieve this by post-multiplying the vector \\(\vec{c}\\) with \\(\mathbf{M}\\), resulting in a \\(2×1\\) matrix:

Reviewer: ```suggestion
In order to multiply a column vector, we must ensure that the number of columns of the matrix on the left hand side matches the number of rows of the matrix on the right hand side. We achieve this by pre-multiplying the vector \\(\vec{c}\\) by \\(\mathbf{M}\\), resulting in a \\(2×1\\) matrix:
```

$$

 The row vector product is as follows:

$$
\begin{bmatrix} 1 & 0 & 1 \end{bmatrix} \cdot \begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix} = \begin{bmatrix} 4 & 0 & 4 \end{bmatrix}
$$

Whereas the column vector product is as follows:

$$
\begin{bmatrix} 1 & 0 & 1 \\\ 0 & 2 & 0 \\\ 3 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 2 \\ 0 \\ 6 \end{bmatrix}
$$

Another way of thinking about this is that in order to have a well-defined product, the row vector has to appear first in the product, and since we have transposed the vector, we also need to transpose the matrix so for \\(\vec{c}\\) we would have:

$$
\begin{array}{l@{}l}
\mathbf{M^T} \cdot \vec{c} 
    &{}= \begin{bmatrix} m_{11} & m_{21} \\\ m_{12} & m_{22}  \end{bmatrix} \cdot \begin{bmatrix} c_{1} \\\ c_{2} \end{bmatrix} \\\ \\\
&{}= \begin{bmatrix} 
m_{11}c_{1} + m_{12}c_{1} \\\ 
m_{21}c_{2} + m_{22}c_{2} \end{bmatrix}
\end{array}
$$

With this in mind, the column vector product between the transpose of the matrix used in the vector product is now equal to that of the vector product as follows:

$$

> \begin{bmatrix} 1 & 0 & 3 \\\ 0 & 2 & 0 \\\ 1 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\\ 0 \\\ 1 \end{bmatrix} =  \begin{bmatrix} 4 \\ 0 \\ 4 \end{bmatrix}

Reviewer: ```suggestion
\begin{bmatrix} 1 & 0 & 3 \\ 0 & 2 & 0 \\ 1 & 0 & 3 \end{bmatrix} \cdot \begin{bmatrix} 1 \\ 0 \\ 1 \end{bmatrix} =  \begin{bmatrix} 4 \\ 0 \\ 4 \end{bmatrix}
```

$$

We can implement the matrix-vector product with column major array storage as follows:

```c++
// 3x1 column vector
const float columnVector[3] = {1.f, 0.f, 1.f};

// 4x3 matrix in column-major order
const std::size_t rows = 4;
const std::size_t cols = 3;

const float columnMajor[cols][rows] = 
{ 
    {1.f, 0.f, 1.f, 4.f}, // column 0 
    {0.f, 2.f, 0.f, 1.f}, // column 1
    {3.f, 0.f, 3.f, 4.f}  // column 2
};

// 4x1 column vector to hold the result
float result[rows] = {0.f, 0.f, 0.f, 0.f};

for(std::size_t r = 0; r < rows; r++)
{
   for(std::size_t c = 0; c < cols; c++)
   {
      result[r] += columnMajor[c][r] * columnVector[c];
   }
}

// prints 4 0 4 8
for(std::size_t i = 0; i < rows; i++)
{
   std::cout << result[i] << " ";
}
```

You may have noticed that the code for both snippets is identical, bar some flipping of variable names for clarity. Remember that the product of a row vector with a given matrix is the transpose of the product between a column vector and the transpose of that same matrix:

$$

> \vec{r} \cdot \mathbf{M} = (\mathbf{M^T} \cdot \vec{c})^\mathbf{T}}

Reviewer: ```suggestion
\vec{r} \cdot \mathbf{M} = (\mathbf{M^T} \cdot \vec{c})^\mathbf{T}
```

$$

An interesting property of storing our matrices in the array matrix order convention of our vector convention (i.e. row vectors with matrices stored in row-major order and column vectors with matrices stored in column-major order) is that the actual array data for the matrices is identical regardless of which of the two convention combinations you choose. Consider the following 2d transform matrix for scaling, anti-clockwise rotating and translating a given row vector:

$$
\begin{bmatrix}      S_{x} \cdot \cos(\theta) & \sin(\theta) & 0  \\\      -\sin(\theta) & S_{y} \cdot \cos(\theta) & 0  \\\  T_{x} &  T_{y}  & 1  \end{bmatrix}
$$

When laid out in row-major in a two dimensional array, the array is initialized as follows:

```c++
// Scale to double and quadruple the x and y axis accordingly
const float Sx = 2.f;
const float Sy = 4.f

// Rotation of approx. 45 degrees (in radians)
const float theta = 0.707;

// Translation of -5 along x axis and 10 along y axis
const float Tx = -5.f;
const float Ty = 10.f;

// 3x3 matrix in row-major order
const std::size_t rows = 3;
const std::size_t cols = 3;

const float rowMajor[rows][cols] = 
{ 

>     {Sx * cos(theta), sin(theta),      0.f}, // row 0 

Reviewer: Fixed 👍

    {-sin(theta),     Sy * cos(theta), 0.f}, // row 1
    {Tx,              Ty,              1.f}  // row 2
};
```

Now consider the same transformation matrix transposed for translating a given column vector:

$$
\begin{bmatrix}      S_{x} \cdot \cos(\theta) & -\sin(\theta) & T_{x} \\\      \sin(\theta) & S_{y} \cdot \cos(\theta) & T_{y} \\\      0 & 0 & 1  \end{bmatrix}
$$

When laid out in column-major in a two dimensional array, the array is initialized as follows:

```c++
// Scale to double and quadruple the x and y axis accordingly
const float Sx = 2.f;
const float Sy = 4.f

// Rotation of approx. 45 degrees (in radians)
const float theta = 0.785;

// Translation of -5 along x axis and 10 along y axis
const float Tx = -5.f;
const float Ty = 10.f;

// 3x3 matrix in column-major order
const std::size_t rows = 3;
const std::size_t cols = 3;

const float columnMajor[cols][rows] = 
{ 

>     {Sx * cos(theta), sin(theta),      0.f}, // column 0 

Reviewer 2: Same comment about the scale in 1,2 and 2,1 elements.

Reviewer: Fixed 👍
